go/src/lib.rs

```diff
@@ -1,9 +1,171 @@
+/// `failure_callback` is the callback that will be called in the case that the Redis command fails.
+///
+/// # Safety
+///
+/// * `connection_request_bytes` must point to `connection_request_len` consecutive properly initialized bytes. It must be a well-formed Protobuf `ConnectionRequest` object. The array must be allocated by the caller and subsequently freed by the caller after this function returns.
+/// * `connection_request_len` must not be greater than the length of the connection request bytes array. It must also not be greater than the max value of a signed pointer-sized integer.
+/// * The `conn_ptr` pointer in the returned `ConnectionResponse` must live until it is passed into `close_client`.
+/// * The `connection_error_message` pointer in the returned `ConnectionResponse` must live until the returned `ConnectionResponse` pointer is passed to `free_connection_response`.
+/// * Both the `success_callback` and `failure_callback` function pointers need to live until the client is closed via `close_client` since they are used when issuing Redis commands.
+// TODO: Consider making this async
+#[no_mangle]
+pub unsafe extern "C" fn create_client(
+    connection_request_bytes: *const u8,
+    connection_request_len: usize,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> *const ConnectionResponse {
+    let request_bytes =
+        unsafe { std::slice::from_raw_parts(connection_request_bytes, connection_request_len) };
+    let response = match create_client_internal(request_bytes, success_callback, failure_callback) {
+        Err(err) => ConnectionResponse {
+            conn_ptr: std::ptr::null(),
+            connection_error_message: CString::into_raw(CString::new(err).unwrap()),
+        },
+        Ok(client) => ConnectionResponse {
+            conn_ptr: Box::into_raw(Box::new(client)) as *const c_void,
+            connection_error_message: std::ptr::null(),
+        },
+    };
+    Box::into_raw(Box::new(response))
+}
+
+/// Closes the given client, deallocating it from the heap.
+///
+/// `client_ptr` is a pointer to the client returned in the `ConnectionResponse` from `create_client`.
+///
+/// This function will do nothing when closing an already closed client (a null client pointer).
+///
+/// # Safety
+///
+/// * `client_ptr` must be obtained from the `ConnectionResponse` returned from `create_client`.
+/// * `client_ptr` must be valid until `close_client` is called.
+// TODO: Ensure safety when command has not completed yet
+#[no_mangle]
+pub unsafe extern "C" fn close_client(client_ptr: *const c_void) {
+    if client_ptr.is_null() {
+        return;
+    }
+    drop(unsafe { Box::from_raw(client_ptr as *mut Client) });
+}
+
+/// Deallocates a `ConnectionResponse`.
+///
+/// This function also frees the contained error. If the contained error is a null pointer, the function returns and nothing happens, so a double free cannot happen.
```

Comment:
how does the contained error pointer being null prevent double free?

Reply:
Ah, when Andrew suggested checking for null to allow `close_client` to be called multiple times, I somehow got the idea that it was a way to avoid double frees. But you're right, this doesn't make any sense.

Reply:
I assume that this comment hints at the go wrapper replacing the client pointer with a null pointer when calling this (which makes sense) - but that's an action on the wrapper's side, not something that the Rust implementation can assume that happens, and so, not something that should be documented here.